### PR TITLE
shadowTerrain, translucent, entity, blockEntity support

### PR DESCRIPTION
**Pull request by IMS212**

### CODE REVIEW

File: src/main/java/net/coderbot/iris/shaderpack/PackDirectives.java [hunk 1]
```diff
@@ -20,40 +20,37 @@
...
 		sunPathRotation = 0.0F;
 		ambientOcclusionLevel = 1.0F;
 		renderTargetDirectives = new PackRenderTargetDirectives(supportedRenderTargets);
-		shadowDirectives = new PackShadowDirectives();
+		shadowDirectives = packShadowDirectives;
 	}
 
 	PackDirectives(Set<Integer> supportedRenderTargets, ShaderProperties properties) {
-		this(supportedRenderTargets);
+		this(supportedRenderTargets, new PackShadowDirectives(properties.getShadowTerrain().orElse(true), properties.getShadowTranslucent().orElse(true), properties.getShadowEntities().orElse(true), properties.getShadowBlockEntities().orElse(true), properties.getShadowCulling()));
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/PackDirectives.java:37]: This line is way too long

File: src/main/java/net/coderbot/iris/shaderpack/PackDirectives.java [hunk 2]
```diff
@@ -20,40 +20,37 @@
...
 		separateAo = properties.getSeparateAo().orElse(false);
 		oldLighting = properties.getOldLighting().orElse(false);
 		explicitFlips = properties.getExplicitFlips();
-		shadowCulling = properties.getShadowCulling();
 		particlesBeforeDeferred = properties.getParticlesBeforeDeferred().orElse(false);
 	}
 
 	PackDirectives(Set<Integer> supportedRenderTargets, PackDirectives directives) {
-		this(supportedRenderTargets);
+		this(supportedRenderTargets, new PackShadowDirectives(directives.getShadowDirectives().shouldRenderTerrain(), directives.getShadowDirectives().shouldRenderTranslucent(), directives.getShadowDirectives().shouldRenderEntities(), directives.getShadowDirectives().shouldRenderBlockEntities(), directives.getShadowDirectives().getCullingState()));
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/shaderpack/PackDirectives.java:49]: Why not make ShadowDirectives have a "copy constructor" here or just pass it as-is?

File: src/main/java/net/coderbot/iris/pipeline/ShadowRenderer.java [hunk 3]
```diff
@@ -667,11 +681,11 @@ public void addDebugText(List<String> messages) {
 		messages.add("[Iris] Shadow Maps: " + debugStringOverall);
 		messages.add("[Iris] Shadow Distance: " + debugStringShadowDistance);
 		messages.add("[Iris] Shadow Culling: " + debugStringShadowCulling);
-		messages.add("[Iris] Shadow Terrain: " + debugStringTerrain);
+		messages.add("[Iris] Shadow Terrain: " + debugStringTerrain + (shouldRenderTerrain ? "" : " (no terrain) ") + (shouldRenderTranslucent ? "" : "(no translucent)"));
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/pipeline/ShadowRenderer.java:684]: Please split this across multiple lines

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Looks good, just some stylistic comments

**Comment by IMS212** [comment]
Fixed

**Review by coderbot16 — APPROVED** [review]
Thanks! Looks good to me

**Pull request merged**